### ISSUE TITLE
Simplify Property type validation

### ISSUE THREAD

**Issue by zepumph** [phetsims/axon#189]
While talking to @samreid about #182, we discussed ways of trying to (a) have the best validation code possible for Emitter and Property as well as (b) have them both share as much api/code as is reasonable.

We came up with a strategy that would allow us to remove `isValidValue` and `validValues`, and instead inline everything into `valueType`

The goal would be to support all types of `valueType` in the same parameter. Here are the pieces we want to support:
* 'number' -> {string} that you can call `typeof value` to get the valueType
* Constructor -> {function} that you can call `value instanceof X` to get the valueType
* predicate functions -> {function} that returns a boolean given the value of if the value is valid.

Property.valueType currently supports the first two, but to support the difference between the second two, @samreid and I have done some brainstorming.

* a constructor has a `name` property. For example:
```
// requirejs mode
phet.scenery.Node.name --> 'Node'

// build mode
phet.scenery.Node.name --> 'e'
```
Anonymous functions don't have a name, so by checking on a name, we could cover many cases, for example:
```js
new Property( { valueType: v => v ===0 || v === null})
```
If there was a case where you wanted a predicate to have a name, then we could make a convention that  that function then needs a reserved property that we can check on:

```js
var namedPredicate = v => { v >=0 || v === null };
namedPredicate.isPredicate = true // or something like that
new Property( { valueType: namedPredicate })
```

The validation would be pretty simple, and not that different than the current form of `AXON/assertValueType`:

```js
  function assertValueType( value, valueType, allowUndefined ) {

. . .
   // ******************************************************************************************
    else if( typeof valueType === 'function' && (!valueType.name || valutType.isPredicate){
       assert( valueType( value) ); // because this is a predicate function
     }
   // ******************************************************************************************
. . .
  }
```

On top of this (sorry for the long issue), having this reserved predicate marker could be really nice for factoring some of these out.

We could have a `Validators` class where we keep some simple validators like
`Validators.greaterThanZero`
`Validators.numberOrNull`
etc. . . 

Each of the above static functions would have the `isPredicate` prop to support the validation as a named function.

We will need to have something like this for each emitter arg as well, so it could be nice to factor it out into a central place

Marking for dev meeting.

**Comment by samreid** [comment]
PhET-iO uses `validValues` to enumerate values for radio buttons, so I'm not sure that would be replaced.  Or maybe we don't need that feature any more?

Also, if we combine all of this checking into one place, we may need a better name than `valueType` (since it would support predicates in general, not just types).

**Comment by pixelzoom** [comment]
Look to lodash for a better name than `valueType`.

**Comment by samreid** [comment]
> Look to lodash for a better name than valueType.

It seems lodash uses the term `predicate` for cases like this.

**Comment by jonathanolson** [comment]
It would be nice to have nice enumeration support here.

What is the recommended way to port validValues?

**Comment by zepumph** [comment]
>What is the recommended way to port validValues?

Depending on how formal we want to get, I like one of the following two ways:

* declare valid values as a function:
```js
predicate: ( value )=>{
   return value === value1 || value === value2 . . .;
}
```

Or we create a static `Validators` type with functions to do this for you:
```js
predicate: Validators.predicateForValidValues( [value1, value2])
```

Maybe the second is a bit clunky, not sure which one I prefer.

**Comment by pixelzoom** [comment]
Validation is not always an equality comparison.  E.g.:

`predicate: value => ( typeof value === 'number' ) && ( value >= 0 )`

`predicate: value => value instanceof Quadratic || value === null`

So validation needs to support general boolean expressions. I don't think the `Validators` approach suggested above is practical for that, and I prefer a function declaration.

**Comment by samreid** [comment]
With our new Enumeration library, we could use:
```js
predicate: MyEnumeration.includes.bind(MyEnumeration)
```

**Comment by pixelzoom** [comment]
If we flatten Property validation into one `predicate` option, here are the things that we need to support.  Some of these are currently supported by Property's validation options, some of these we've had to workaround by writing `isValidValue` functions.

* `value => {{boolean expression}}`
* `instanceof` one or more classes
* `typeof` one or more JS types
* `Enumeration`
* `null`
* combinations of some of the above?

**Comment by zepumph** [comment]
This was discussed in the meeting today. A large focus was trying to see how the phet-io needs would fit into a general case.

There seemed to be a consensus that moving to a pattern that only accepted functions as predicates seemed ok. This goes against what @pixelzoom recommended above, and instead looked at converting all of the above list into functions to be passed in as the predicate option (whereas before they were "converted" into predicate functions in Property. 

Then the `Validator` could server to factor out much of this logic:
* instanceof --> `Validator,instanceof( Node )`
* typeof --> `Validator.typeof ( 'number')` or `Validator.number()` (and one function for each typeof call possible.
* Enumeration --> `Validator.validValues()`
* null --> `Validator.null()` and perhaps also  `Validator.nullOr( Validator.number() )`
* combinations of the above. --> Not sure how this would work!

While the predicate option would accept an arbitrary anonymous function, for PhET-iO support it would be important to use the Validator options so that they could send along metadata to the wrapper frame about what the function is doing. 

A main con of this to me is the verbosity. Instead of having the option to just pass `valueType: 'number'` and be done with it, the predicate option would be more verbose. This is less important in Property than it is in Emitter, which can have arbitrary numbers of args, and something like:
`valueTypes: ['number', 'string', Node]`
turns into
`valueTypes: [Validator.number(), Validator.string(), Validator.instanceof( Node )]`

That doesn't look very appealing to me, especially because for simple cases like above (where we aren't testing complex opperations like "vale is null or less -4") we likely can get at the PhET-iO metadata from other places.

We decided to talk about it more next week.

**Comment by pixelzoom** [comment]
@samreid said:
> ... This goes against what @pixelzoom recommended above ...

To clarify, my recommendation was based on the requirements that @zepumph established in the first comment of this issue:
> The goal would be to support all types of valueType in the same parameter. 

If you want to change the requirements and convert everything to predicate functions, I might have a different recommendation.

**Comment by pixelzoom** [comment]
See also https://github.com/phetsims/axon/issues/195, proper handling of `valueType: Array`.

**Comment by zepumph** [comment]
Marking as high priority so that we get to it next dev meeting. A fair bit of work for emitter and property is blocked by this issue.

**Comment by zepumph** [comment]
Again this was discussed again today. 

Most of the conversation was taken up in just understanding the current limitations, as well as the variety of potential solutions that we could go with.

Right now we are basically trying to determine a general, and consistent way to conduct value validation. In Property we want to have a value validated in the same way as for each argument for Emitter, as well as testing values in TypeIOs.

At one point, there was a thought to make generic value validation only ever accept a `function`. So each arg in Emitter would need a corresponding  function (predicate) that when given a potential argument value, the function returned whether that value was valid.

There was some consensus that it is really nice to be able to specify simple typeof strings and Type constructors as well as predicate functions. I.e basically what `Property`'s `valueType` option supports right now.


At one point there was talk of having an option in Emitter to support the arg value validations, and then when instrumenting this Emitter, to convert those over to where the `phetioType` was created. @ariel-phet felt strongly that we shouldn't have to touch this code once to add phet validation, and again to convert it to phet-io, since it is basically just copying functionality over from one option to another place.

So for next week. I'm going to implement a solution that involves just using `phetioType` for validation to see how it goes. This means that even uninstrumented Emitters would use `phetioType` to validate the emitter args.

Note that @pixelzoom expressed concern in how we are conflating the phetioType with the accepted value validation, which seemed weird. We weren't able to finish that piece of the discussion before the dev meeting ended.

**Comment by samreid** [comment]
Pushed to next week.

**Comment by zepumph** [comment]
I'm sorry I didn't get to this again. I will work on it this afternoon and we will talk next week.

**Comment by zepumph** [comment]
I went down a road after a lot of thought from the above documented conversation, and by looking at many potential usages of Emitter, both for phet-io and phet. 

Proposal implemented entirely in the above commit:

I formalized the definition of what is considered a "type" that a value can be checked against and created `TypeDef.js`. Then I updated Emitter to test it out.

All Emitter arg validation to be done through EmitterIO.
What was 
`    this.finishEmitter = new Emitter( { valueTypes: [ 'number' ] } );`
 is now
`    this.finishEmitter = new Emitter( { phetioType: EmitterIO( [ { type: 'number' } ] ) } );`

`type` property is of Type {TypeDef} that is meant to be flexible, as a string, constructor, predicate function, or TypeIO (which has its own default type predicate. 

`type` is required for each argument in the EmitterIO constructor

For phet-io support we also added `predicate` key so that you can have the `type` be a typeIO, and the `predicate` a more specific function. This is in case the default IO type predicate is not specific enough. Example:
`{ type: NumberIO, predicate: v => typeof v === 'number' && v > 5 }`



**Best places to look at example usages:**
* EmitterTests.js
* Atom2.js
* ariaHerald.js
* PressListener.js
* Animation.js

Example "validators" at the end of `TypeDef.js`

**Pros:**
I really like TypeDef.js I think we have needed something like it for a while. It is consistent and can be applied to many different use cases.

**Problems:**
* Basically all predicate functions need the `isPredicate` property. the naming hack doesn't seem to work.
* Lots more boiler plate (`valueTypes: [ 'string', 'boolean' ]` --> `phetioType: EmitterIO( [ { type: 'string' }, { type: 'boolean' } ] )`)
* Up for discussion, is TypeDef.js to convoluted, and we shouldn't try to pack everything into a single type.


We can revert this and go with something else, but I thought it would be good to at least start implementing something, so that we can better discuss next week at dev meeting.

**Comment by samreid** [comment]
Some preliminary thoughts:

```js
// Option A
predicate: node => node === null || node instanceof Node }
```

seems preferable to:
```js
// Option B
predicate: TypeDef.getNullOrInstanceOfPredicate( Node )
```
because Option A is simple very clear what is happening without needing to rely on a library function.  But Option A is not marked as `isPredicate = true` so, would it work?  Also, I'm unclear on why things need to be marked as `isPredicate` or not, especially if they are put as a `predicate` attribute.  Presumably that is for another part of type checking.

I do see the advantage of factoring out things like `nullOrFunctionPredicate`, especially if they are prevalent.

**Comment by zepumph** [comment]
Unfortunately the `isPredicate` flag is the only reliable way for `TypeDef` to know if the function "type" it has is like `Node`, and `instanceof` should be checked on it, or if it is a predicate, and it should be run and a boolean expected.

You are correct that in this case, where we are passing it in as a `predicate` value to EmitterIO, that we could get around this, but I was also trying to support this in the `type` key, as well as Property usages and anywhere that we want to support the generic `TypeDef` in the future.

**Comment by samreid** [comment]
Got it, thanks!

**Comment by zepumph** [comment]
I will come back next week with a discrete list of options to choose from.

**Comment by zepumph** [comment]
I have hashed out 5 options. I think we will likely come up with a combination of some of them, but it is good to brainstorm. 


* Option 1: 
  adapt Property and Emitter to use the {TypeDef} type when defining an allowed value/argument type. Usages are different between phet and 
  phet-io instrumented instances.

  PhET-iO Uninstrumented:

  Property:
    ```js
  new Property( 0, valueType: 'number');
  new Property( 0, valueType: value=> value === 0);
  new Property( 'string', valueType: 'number');
    ```
  Emitter:
  ```js
  new Emitter( { // emitter can only emit 3 arguments, and the first must be number, the second must be Node, and the third must be the number 0
    argumentTypes: [ 'number', Node, value=>value===0]
  });

  new Emitter(); // no args emitted
  new Emitter( { // emitter can only emit 3 arguments, and the first and second must be Node, and the third a number
    arguementTypes: [Node, Node, 'number'] 
    })
  ```

  PhET-iO Instrumented:  

    Property:
    ```js
  new Property( 0, phetioType: NumberIO);
  new Property( 0, valueType: value=> value === 0);
  new Property( 'string', valueType: 'number');
    ```
  Emitter:
  ```js
  new Emitter( { // emitter can only emit 3 arguments, and the first must be nubmer, the second must be Node, and the third must be the number 0
    argumentTypes: [ 'number', Node, value=>value===0]
  });

  new Emitter(); // no args emitted
  new Emitter( { // emitter can only emit 3 arguments, and the first and second must be Node, and the third a number
    arguementTypes: [Node, Node, 'number'] 
    })
  ```



* Option 2: formulate the PhetioType from given "sub" TypeIOs, which come from the value (Property) or args (Emitter):
```js
// then property knows that this is a TypeIO, and automatically creates the PropertyIO() phetioType from that
  new Property( 0, valueType: NumberIO); 

  // phetioType becomes PropertyIO( StringIO )
  new Property( 'hello', valueType: StringIO); 

  // phetioType becomes PropertyIO( NullableIO( StringIO) )
  new Property( 'hello', valueType: NullableIO( StringIO) ) 

  // since this is using a predicate, we need to specify a phetioType too for instrumentation
  new Property( 'hello', valueType: value=> value.startsWith( 'h'), phetioType: PropertyIO(String)) -> 


  // still need to think about EmitterIO args supporting doc and naming
  new Emitter( {
    argumentTypes: [NumberIO, StringIO, NullableIO(StringIO) ] // this emitter would work for phet-io out of the box
    })

  new Emitter( {
    argumentTypes: [value=> x >=3 ] // not sure how exactly to support this one

  });  
```


* Option 3:
  We don't use typeDef, Property stays the same, and emitter supports the three options that Property uses but for each arg. Not sure exactly what that would look like:
  ```js

  new Emitter( {
    agrumentTypes: [ { valueType: 'number'}, {isValidValue: value=> x>=3}, {validValues: [1,2,5,9]}]
    });
   ```
   I'm not sure yet what PhET-iO support would look like for that


* Option 4:
  Logic in Emitter and Property diverge, so maybe Property stays the same with these three options, but Emitter gets a revamp with TypeDef



* Options 5:
  all type validation happens through phetioType
  Here you will have to specify a phetioType EVEN IF it isn't instrumented
  Pro is that instrumenting an instance becomes that much easier.


```js

// nullable and numberIO know how to validate this value
new Emitter( phetioType: EmitterIO([ { type: NullableIO(NumberIO)) }])}]);

// here we can see that the first arg accepts a predicate instead of an IOType 
// (note to self: I wonder if can structured cloning help us here so we don't need an IOType?)
new Emitter( phetioType: EmitterIO([ { type: v=> typeof v === 'number' || v === null  }])}]);

// No need for valueType, we get that from NumberIO
new Property( {  phetioType: Property( NumberIO ) });

new Emitter( { 
phetioType: EmitterIO( [NumberIO, StringIO, BlargIO]) // all of the argument types are gotten from the EmitterIO's argumentTypes
 });
  ```

**Comment by samreid** [comment]
One way to support PhET-iO for Option 3.  This matches precedent set by Property.js and doesn't force us into providing parallel arrays (like if we supplied argumentTypes alongside phetioType: EmitterIO(ioTypeArray)).  Main disadvantage is that it breaks the pattern where you pass `phetioType` to an instantiation.

```js
new Emitter( {
  argumentTypes: [ {
    name: 'point',
    valueType: Vector2,
    phetioType: Vector2IO,
    documentation: 'point in view coordinates'
  }, {
    name: 'age',
    isValidValue: value => x >= 3,
    phetioType: NumberIO,
    documentation: 'age in years'
  }, {
    name: 'heights',
    validValues: [ 1, 2, 5, 9 ],
    phetioType: ArrayIO( NumberIO ),
    documentation: 'heights in feets'
  } ]
} );
```

**Comment by zepumph** [comment]
Notes from discussion with @samreid about how to proceed.


```js
// This is great! For phetio brand
new Emitter( {
  phetioType: EmitterIO( [ {
    name: 'point',
    phetioType: Vector2IO, // value type is retrieved from Vector2IO (Vector2IO.type === Vector2)
    documentation: 'point in view coordinates'
  }, {
    name: 'age',
    isValidValue: value => x >= 3,
    phetioType: NumberIO,
    documentation: 'age in years'
  }, {
    name: 'heights',
    validValues: [ 1, 2, 5, 9 ],
    phetioType: NumberIO,
    documentation: 'heights in feets'
  } ] )
} );


// phet brand, uninstrumented
new Emitter( {
  argumentTypes: [ {

    // either one of these, not both
    valueType: Vector2,
    phetioType: Vector2IO, // this is easier to then later convert to phet-io, but requires the knowledge of phet-io to do it
  }, {
    isValidValue: value => x >= 3
  }, {
    validValues: [ 1, 2, 5, 9 ]
  } ]
} );


// phet brand

new Property( new Vector2( 2, 3 ), {
  valueType: Vector2,
  tandem: tandem.createTandem( 'aVector2')
} );


new Property( new Vector2( 2, 3 ), {
  phetioType: PropertyIO( NullableIO( Vector2IO ) ), // NOT BOTH, just phetioType here!!!! But it still works in phet brand
  tandem: tandem.createTandem( 'helloVector2OrNullasAValueType')
} );

// phet brand
new Property( new Vector2( 2, 3 ), {
  valueType: Vector2, // or
  phetioType: PropertyIO( Vector2IO )// NOT BOTH!!!! 
} );

```
  

NOTES:
* For emitter, you can't provide phetioType AND argumentTypes. In addition, each entry cannot have a `valueType` and `phetioType`, because they fill the same niche (I think).
* For Property, Cannot provide `phetioType` and `valueType`, assertion
* No reference from Constructor --> IOType, because that is too clever, only reference from IOType -> Constructor.
* Property would be as it is now, but adding the phetioType as another potential validation strategy
* TypeDef to support Type validation via all 4 options. 
    * main difference: explicit keys instead of checking on the "value" of type
    * "validValue" would accept an Object of options with a value, would assert out if none of the 4 keys were provided. Otherwise it would be a boolean. No assertions in the `validValue` function
    * would know the precedent of all 4 keys, would know how to determine if something was a valid value (like Property is already basically doing, just copy that over).
    * a reusable library can have unit tests built around it. As well as being used by any file that wants it.
    * logic in Proprety is tried and true, we should utilize all of it in the "TypeDef" library, and add on phetioType.
    * A key to this not being very complicated is that phetioType is just a standin/replacement for valueType. It isn't trying to do validValues or valueType.
    * (extra credit) For phetio, perhaps add support for validValues being serialized into metadata for more than just Property, in some reusable kind of way instead of `Property.toStateObject`.


Here is some pseudo code that took some logic from Property. It would be adapted to output boolean instead of asserting, but many assertions about how different options can act together would stick around.
```js
const TypeDef = {

  /**
   * Test if a value is valid
   * @param {*} value
   * @param {Object} options
   * @returns {boolean} - if the value is a valid one given the valueType
   */
  validValue( value, options ) {
    // add assert it has some of the 4 options

    // validate options
    // assertions from Property
    // assert that you have EITHER phetioType or valueType


    // simlar asserting function as in Property, but instead we likely want to return booleans
    this.assertPropertyValidateValue = assert && function( value ) {

      // this changes to getting the type from either the phetioType or the valueType
      options.valueType && assertValueType( value, options.valueType );

      options.validValues && assert( options.validValues.indexOf( value ) !== -1,
        'value is not a member of validValues: ' + value );

      options.isValidValue && assert( options.isValidValue( value ),
        'value failed isValidValue test: ' + value );
    };

    // verify that validValues meet other validation criteria
    if ( this.assertPropertyValidateValue && options.validValues ) {
      options.validValues.forEach( function( value ) {
        self.assertPropertyValidateValue( value );
      } );
    }

    // validate the initial value
    this.assertPropertyValidateValue && this.assertPropertyValidateValue( value );


  },
};
 ```

Steps to proceed: 
  * Factor out Property type validation into it's own file now! To master. That's a step in the right direction! We don't need to add any extra support to it right now.
  * Then get PropertyIO working with that. . . (pending another dev meeting potentialy?)
  * Unit tests for this validation library
  * Emitter would be next.
  * Add support for phetioType as an option too?

**Comment by samreid** [comment]
I've seen a few places where this type validation could be useful outside of Emitter and Property. For instance, this code:

```js
assert && assert( config.numberOfSources === 1 || config.numberOfSources === 2, 'Must have 1 or 2 sources' );
```

Could become:
```js
assert && TypeDef.assert( config.numberOfSources, { validValues: [ 1, 2 ] }, 'Must have 1 or 2 sources' );
```

Also, `TypeDef` may not be the best name for this type.  Perhaps something like `Validator` would be better.

**Comment by samreid** [comment]
`Validator` seems promising.  `Type.*` is not accurate because it is not confined to checking types.  It is unfortunate that `isValidValue` is already reserved, because it is the perfect name for the `Validator` method.  I agree about the 2 methods you identified.

**Comment by samreid** [comment]
This would come in handy for https://github.com/phetsims/wave-interference/issues/275, I may try out some ideas.

**Comment by samreid** [comment]
Update: I've made significant progress in factoring out validation from Property and reusing it for options.  Not to a commit point yet, but wanted to let you know I'm headed that way.

**Comment by samreid** [comment]
A brief summary of the work so far:

* Added Validator.js, which is a singleton described as: Throws assertion errors if a value doesn't match the specified criteria.
* Refactored Property.js to use Validator.js.  This simplifies the logic in Property.js and makes its same logic reusable elsewhere.  I added features to Validator to keep reasonable performance for Property--this was important because Property uses the same validation options many times.
* Added ValidatorTests.  These are very brief sanity tests--most of the functionality is still covered in PropertyTests.
* inlined assertValueType into Validator.js.  It simplified things and it no longer seems that assertValueType needs standalone usage
* phetioInherit no longer needs to attach `isPredicate` to `isInstance`.  I'm not 100% sure about this part.
* Added Validator usage in WavesModel for https://github.com/phetsims/wave-interference/issues/275.  This helps to keep the config validation consolidated and we may want to use this pattern in other options/config validation.

I decided to commit to master since the points above generally seem like improvements, because it will make it simplest to coordinate reviews and discussion and moving forward, and to get additional testing on Bayes CT.  The changes span 3 repos.
Before committing, I verified:
* lint everything is passing
* Unit tests are passing, except for a failure in Balloons and Static Electricity which looks unrelated
* Aqua fuzz phet brand is passing
* Aqua fuzz phet-io brand is passing, except for pre-existing `Node.offStatic was called but no listener was removed` and `This PhetioObject has already been disposed, and cannot be disposed again` issues.

Next steps:
* Code review.  @pixelzoom would be a natural fit, in part because this overlaps with https://github.com/phetsims/wave-interference/issues/275. @zepumph would be good to review the overall usage in Property and how it can be used in Emitter.
* Note the TODO: now that we have Validator, do we still want to overload valueType?  Or split it up into: `{typeOf: 'boolean'}` vs `{instanceOf: 'Vector2'}` (not sure about the best key names here).
* Leverage this code for Emitter

**Comment by samreid** [comment]
Changes pushed, I'll mark this as blocks-publication until it's been reviewed.  I'll also keep an eye on Bayes CT.

**Comment by samreid** [comment]
Here's a patch that demonstrates one API for consolidating validation and option defaults, using Slider.options.orientation as an example:

<details>

```
Index: js/Slider.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/Slider.js	(revision 1d02acbb9d296a39c65ccf7ea0f873099985bbde)
+++ js/Slider.js	(date 1545501141000)
@@ -33,6 +33,7 @@
   var sun = require( 'SUN/sun' );
   var Tandem = require( 'TANDEM/Tandem' );
   var Util = require( 'DOT/Util' );
+  var Validator = require( 'AXON/Validator' );
 
   // constants
   var VERTICAL_ROTATION = -Math.PI / 2;
@@ -49,8 +50,6 @@
 
     options = _.extend( {
 
-      orientation: 'horizontal', // 'horizontal'|'vertical'
-
       // track
       trackSize: new Dimension2( 100, 5 ),
       trackFillEnabled: 'white',
@@ -102,14 +101,13 @@
       phetioType: SliderIO
     }, options );
 
+    // @private
+    this.orientation = Validator.option( options.orientation, 'horizontal', { validValues: [ 'horizontal', 'vertical' ] } );
+
     assert && assert( range instanceof Range, 'range must be of type Range:' + range );
-    assert && assert( options.orientation === 'horizontal' || options.orientation === 'vertical',
-      'invalid orientation: ' + options.orientation );
     assert && assert( !( options.enabledProperty && options.enabledPropertyOptions ),
       'enabledProperty and enabledPropertyOptions are mutually exclusive' );
 
-    this.orientation = options.orientation; // @private
-
     Node.call( this );
 
     var ownsEnabledProperty = !options.enabledProperty;
@@ -197,7 +195,7 @@
     // This allows us to still decorate the Slider with additional children.
     // See https://github.com/phetsims/sun/issues/406
     var sliderPartsNode = new Node( { children: sliderParts } );
-    if ( options.orientation === 'vertical' ) {
+    if ( this.orientation === 'vertical' ) {
       sliderPartsNode.rotation = VERTICAL_ROTATION;
     }
     this.addChild( sliderPartsNode );
@@ -308,7 +306,7 @@
     // mix accessible slider functionality into Slider
     this.initializeAccessibleSlider( valueProperty, this.enabledRangeProperty, this.enabledProperty,
       _.extend( {}, options, {
-        ariaOrientation: options.orientation
+        ariaOrientation: this.orientation
       } ) );
 
     // support for binder documentation, stripped out in builds and only runs when ?binder is specified

```
</details>

If desired, we could also add a lazy variant so that default values aren't instantiated unless they are needed.  This has been a problem on a number of occasions, here's one API to solve that (though note in this case the overhead of creating a string is not enough to warrant the lazy approach--it would be more for heavier objects, or Properties or instrumented instances.)

```js
// @private
this.orientation = Validator.optionLazy( options.orientation, () => 'horizontal', {
  validValues: [ 'horizontal', 'vertical' ]
} );
```

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/axon/issues/189#issuecomment-449582254, @samreid said:
> Added Validator usage in WavesModel for phetsims/wave-interference#275. This helps to keep the config validation consolidated and we may want to use this pattern in other options/config validation.

Validator consolidates the description, validation and assignment of each config field.  But it in no way provides a consolidated view of the entire config.  Imo, for options and config, Validator is not a superior pattern.  See https://github.com/phetsims/wave-interference/issues/275#issuecomment-449589365

**Comment by pixelzoom** [comment]
In said https://github.com/phetsims/axon/issues/189#issuecomment-449587801, @samreid said:
> Here's a patch that demonstrates one API for consolidating validation and option defaults, using Slider.options.orientation as an example:

Please don't apply this to common code until we've had a dev discussion.  I'm not at all in agreement that Validator is a good replacement for the current `options` pattern.

**Comment by zepumph** [comment]
@samreid @chrisklus and @zepumph looked into `validateOptionsOnValidateValue` today. @samreid noted that his purpose for creating it was to increase performance in the `set` call, so that Property can opt out of calling `validateOptions` each time you call `Property.set()`. We calculated that with validating options  was 240% the time that it took to just `validate`, i.e on average in SOM validate took 5ms a second, and with option validation took 12ms every second. We are glad to have collected this data, but still don't know if it is worth the added complexity, since it is only in requirejs mode anyways (hidden behind assertions).

**Comment by zepumph** [comment]
Right! That is what we were questioning. I think right now I am leaning toward getting rid of the complexity (for example in Property's option extend call) because this hit is only done during requirejs mode.

On the other hand. We know that @jonathanolson has implemented some tests/checks under `assertSlow` for this exact reason. I don't think we are talking about performance problems on that level, but anything being called upwards of 30,000 times a second like `Property.set` during fuzzing I think deserves some discussion.

**Comment by pixelzoom** [comment]
From Property.js -

```js
      // By default, check the options once in the constructor, not on each subsequent value validation, to improve
      // performance in requirejs mode
      validateOptionsOnValidateValue: false
```

The default is not to validate?... We might as well not validate anything if this is the default.

**Comment by zepumph** [comment]
I thought this too! But notice that property is validating options in the constructor still. Then @samreid showed me the possibilities for generality with usages in wave interference. After looking there,  how do you feel about the option?

**Comment by pixelzoom** [comment]
> I thought this too! But notice that property is validating options in the constructor still. 

Property's constructor is validating initialization only.  That's better than nothing, but close to nothing, and certainly not an appropriate default when assertions are enabled.

> Then @samreid showed me the possibilities for generality with usages in wave interference. After looking there, how do you feel about the option?

What am I looking at in Wave Interference?  I see that Scene is using Validator for some validation that's not related to Property, and that's nice. But what justification does that provide for not validating Property after initialization?

**Comment by pixelzoom** [comment]
OK... `validateOptionsOnValidateValue` is confusing.  Next question is, if validation options are indeed immutable, why is `validateOptionsOnValidateValue` needed at all?  When would validation of validation options need to be done more than once?

**Comment by pixelzoom** [comment]
Couldn't `Validation.pluckOptions` validate what it plucks?  

Also recommend to rename `pluckOptions` to `pickOptions`.  Options are picked or omitted, not plucked.

**Comment by samreid** [comment]
> OK... `validateOptionsOnValidateValue` is confusing. 

I was worried it would be confusing.  I considered omitting it altogether but wasn't ready to incur an up to 1% `?ea` performance penalty without discussing it with the team, so I have instead attempted to document it thoroughly and I'm glad we are discussing it.

> Next question is, if validation options are indeed immutable, why is `validateOptionsOnValidateValue` needed at all?

There are two types of use cases of Validator.validate().  Type A (exemplified in Property and will also be seen in Emitter) is called repeatedly with the same options but different values.  Type B is called only once, like this example in Wave Interference's Scene.js:

```js
// @public (read-only) {number}
this.numberOfSources = Validator.validate( config.numberOfSources, { validValues: [ 1, 2 ] } );
```

In both cases, it seems appropriate to validate the validation options themselves, such as making sure the user didn't pass a non-array like `{validValues: 'hello'} `.  However, for Type A, we can optimize by only validating the values on the first pass and never again. We could alternatively solve this with a separate method on Validator's interface.  Or we could think of a better name than `validateOptionsOnValidateValue`?  Or... maybe Validator could "stamp" approved options so this would be handled automatically and not appear in the API?  But that seems kind of magical and may lead to worse confusion?

> Couldn't `Validation.pluckOptions` validate what it "plucks"?  

Sure, but `pluckOptions` is never called for Type B so that doesn't solve the root of the problem.

> Also recommend to rename `pluckOptions` to `pickOptions`.  Options are picked or omitted, not plucked.

Thanks, I'll rename that.

**Comment by samreid** [comment]
> Or... maybe Validator could "stamp" approved options so this would be handled automatically and not appear in the API?

Here's a patch that accomplishes that.  Once a set of options has passed validation, it is marked as approved.  Options marked as approved aren't checked again.  However, this has the side effect that it mutates options.  Or if options are mutated and reused it could have an incorrect approval, etc. etc.

```
Index: js/Property.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/Property.js	(revision 0df6e306ce59eac4851dd207a01095668801936f)
+++ js/Property.js	(date 1545882767000)
@@ -44,15 +44,12 @@
       // Use this to detect or prevent update cycles. Update cycles may be due to floating point error,
       // faulty logic, etc. This may be of particular interest for PhET-iO instrumentation, where such
       // cycles may pollute the data stream. See https://github.com/phetsims/axon/issues/179
-      reentrant: false
+        reentrant: false
+      },
 
       // See Validator.DEFAULT_OPTIONS for validation options.
-    }, Validator.DEFAULT_OPTIONS, {
-
-      // By default, check the options once in the constructor, not on each subsequent value validation, to improve
-      // performance in requirejs mode
-      validateOptionsOnValidateValue: false
-    }, options );
+      Validator.DEFAULT_OPTIONS,
+      options );
 
     // @private
     this.validatorOptions = Validator.pickOptions( options );
Index: js/Validator.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/Validator.js	(revision 0df6e306ce59eac4851dd207a01095668801936f)
+++ js/Validator.js	(date 1545882656000)
@@ -36,11 +36,7 @@
     // {function|null} function that validates the value. Single argument is the value, returns boolean. Unused if null.
     // Example:
     // isValidValue: function( value ) { return Util.isInteger( value ) && value >= 0; }
-    isValidValue: null,
-
-    // By default, always check the provided options.  Can be turned off for cases where the same options are used
-    // repeatedly, such as in Property
-    validateOptionsOnValidateValue: true
+    isValidValue: null
   };
 
   const DEFAULT_OPTIONS_KEYS = _.keys( DEFAULT_OPTIONS );
@@ -58,8 +54,7 @@
 
       if ( assert ) {
 
-        // validate options, but allow opting out for improved performance for repeat cases like Property
-        options.validateOptionsOnValidateValue && Validator.validateOptions( options );
+        Validator.validateOptions( options );
 
         // TODO: now that we have Validator, do we still want to overload valueType?  Or split it up into:
         // TODO: {valueTypeOf: 'boolean'} vs {valueInstanceOf: 'Vector2'}
@@ -91,7 +86,7 @@
      * @public
      */
     validateOptions: options => {
-      if ( assert ) {
+      if ( assert || !options.validatorApproved ) {
         const valueType = options.valueType;
         assert(
           typeof valueType === 'function' ||
@@ -122,6 +117,7 @@
           const remainingOptions = _.omit( options, 'validValues' );
           options.validValues.forEach( v => Validator.validate( v, remainingOptions ) );
         }
+        options.validatorApproved = true;
       }
       else {
         return true;
```

**Comment by samreid** [comment]
This issue overlaps somewhat with https://github.com/phetsims/wave-interference/issues/275#issuecomment-450068968

**Comment by samreid** [comment]
From today's dev meeting:

Conversation A: using Validator in Property, and in standalone and in Emitter
CM: Seems OK.  
	SR: MK and I will leverage that in Emitter.

	CM: can NumberProperty use validator for 

	    // @private {function|null} value validation that is specific to NumberProperty, null if assertions are disabled
    this.assertNumberPropertyValidateValue = assert && function( value ) {
      if ( options.numberType === 'Integer' ) {
        assert( value % 1 === 0, 'numberType was Integer but value was ' + value );
      }
      options.range && assert( value >= options.range.min && value <= options.range.max,
        'value is out of range, value=' + value + ', range=[' + options.range.min + ',' + options.range.max + ']' );
    };

**Comment by zepumph** [comment]
Previous list of "next step"
>Code review. @pixelzoom would be a natural fit, in part because this overlaps with phetsims/wave-interference#275. @zepumph would be good to review the overall usage in Property and how it can be used in Emitter, and double checking that the change in phetioInherit is good.

This change in phetioInherit looks good. Right now we aren't using `isInstance` at all, but may with changes in #204.

>Note the TODO: now that we have Validator, do we still want to overload valueType? Or split it up into: {typeOf: 'boolean'} vs {instanceOf: 'Vector2'} (not sure about the best key names here).

I feel like it is pretty safe to differentiate these 2 based on checking `type X === 'string'`. If it isn't then test instanceof. 

>Leverage this code for Emitter
Out for review as part of #182 

>Can this approach replace our typical options pattern, so that everything about an option is consolidated? (We would need an alternative to _.extend).
see #202 

Next steps for this issue:
- [ ] discuss "pickOptions" pattern, and if we should keep it.
- [ ] Implement Validator logic in NumberProperty

@samreid can you think of anything else to do here that won't be covered by follow-up issues?

**Comment by samreid** [comment]
> I feel like it is pretty safe to differentiate these 2 based on checking `type X === 'string'`. If it isn't then test instanceof. 

Sounds reasonable, I'll remove the corresponding TODO from the code.

**Comment by pixelzoom** [comment]
@zepumph said:
> I feel like it is pretty safe to differentiate these 2 based on checking type X === 'string'. If it isn't then test instanceof.

@samreid said:
> Sounds reasonable, I'll remove the corresponding TODO from the code.

If I understand correctly, the proposal is to make `valueType` support both instanceof and typeof checks. For example: `valueType: Vector2` and `valueType: 'string'`.   If that's the case... Reminder that `Array` needs to be handled specially, see https://github.com/phetsims/axon/issues/195.

**Comment by samreid** [comment]
> If I understand correctly, the proposal is to make `valueType` support both instanceof and typeof checks.

Not just a proposal, that is how Property validation has been since it was added in February 2018 in 866c63d7165bf019dddc59189ea9fa5f54ebc0af

I agree those are unsuitable for `Array`, we would probably need to use `isValidValue` or build custom support for it.

**Comment by zepumph** [comment]
>validateOptionsOnValidateValue

very good thanks!

Good to note that https://github.com/phetsims/axon/issues/195 will be worked on more, but this feels like things have wrapped up well. 

I'm going to close, and new issues can be created from here. Thanks all!
